### PR TITLE
Board support for nRF52832 and nRF52840 from makerdiary

### PR DESCRIPTION
**Pull request by jpsaman**
Merging zephyrproject-rtos/zephyr with makerdiary/zephyr adds board support for nRF52832-mdk and nRF52840-mdk.

### CODE REVIEW

File: boards/arm/nrf52832_mdk/Kconfig.defconfig [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_NRF52832_MDK
+
+config BOARD
+	default "nrf52832_mdk"
+
+if UART_NRFX
+
+config UART_0_NRF_TX_PIN
```
Comment (galak) [inline boards/arm/nrf52832_mdk/Kconfig.defconfig]: uart pins should come from dts.

File: boards/arm/nrf52832_mdk/board.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __INC_BOARD_H
+#define __INC_BOARD_H
+
+#include <soc.h>
+
+/* Push button switch 0 */
```
Comment (galak) [inline boards/arm/nrf52832_mdk/board.h]: button and led should come from dts

File: boards/arm/nrf52840_mdk/Kconfig.defconfig [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_NRF52840_MDK
+
+config BOARD
+	default "nrf52840_mdk"
+
+if UART_NRFX
+
+config UART_0_NRF_TX_PIN
```
Comment (galak) [inline boards/arm/nrf52840_mdk/Kconfig.defconfig]: uart pins should come from dts.

File: boards/arm/nrf52840_mdk/board.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __INC_BOARD_H
+#define __INC_BOARD_H
+
+#include <soc.h>
+
+/* Push button switch 0 */
```
Comment (galak) [inline boards/arm/nrf52840_mdk/board.h]: button & leds should come from dts

File: soc/arm/nordic_nrf/nrf52/dts_fixup.h [hunk 5]
```diff
@@ -26,8 +26,16 @@
 #define CONFIG_UART_0_IRQ_NUM		NORDIC_NRF_UART_40002000_IRQ_0
 #define CONFIG_UART_0_BAUD_RATE		NORDIC_NRF_UART_40002000_CURRENT_SPEED
 #define CONFIG_UART_0_NAME		NORDIC_NRF_UART_40002000_LABEL
-#define CONFIG_UART_0_TX_PIN		NORDIC_NRF_UART_40002000_TX_PIN
-#define CONFIG_UART_0_RX_PIN		NORDIC_NRF_UART_40002000_RX_PIN
+  #if defined(NORDIC_NRF_UART_40002000_TX_PIN)
```
Comment (galak) [inline soc/arm/nordic_nrf/nrf52/dts_fixup.h]: nack, move the pin cfg to dts, than this isn't needed.

File: samples/basic/blink_led/src/main.c [hunk 6]
```diff
@@ -36,6 +36,9 @@
 #elif defined(CONFIG_BOARD_COLIBRI_IMX7D_M4)
 #define PWM_DRIVER	PWM_1_LABEL
 #define PWM_CHANNEL	0
+#elif defined(CONFIG_SOC_FAMILY_NRF)
```
Comment (galak) [inline samples/basic/blink_led/src/main.c]: need to add the platforms to the whitelist in samples/basic/blink_led/sample.yaml

File: dts/arm/nordic/nrf52840.dtsi [hunk 7]
```diff
@@ -1,3 +1,4 @@
+#include <mem.h>
```
Comment (galak) [inline dts/arm/nordic/nrf52840.dtsi]: nack, you shouldn't be modifying nrf52840.dtsi, probably need to include nrf52840_qiaa.dtsi instead.

File: boards/arm/nrf52840_mdk/nrf52840_mdk.dts [hunk 8]
```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/dts-v1/;
+#include <nordic/nrf52840.dtsi>
```
Comment (galak) [inline boards/arm/nrf52840_mdk/nrf52840_mdk.dts]: change to include <nordic/nrf52840_qiaa.dts>

File: boards/arm/nrf52840_mdk/Kconfig.defconfig [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+config UART_0_NRF_RTS_PIN
+	int
+	default 5
+
+config UART_0_NRF_CTS_PIN
+	int
+	default 7
+
+if MODEM_WNCM14A2A
```
Comment (galak) [inline boards/arm/nrf52840_mdk/Kconfig.defconfig]: do you need MODEM_WNCM14A2A?

File: boards/arm/nrf52832_mdk/nrf52832_mdk.dts [hunk 10]
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2018 Zelin <zelin@makerdiary.com>
+ * Copyright (c) 2018 makerdiary.com.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/dts-v1/;
+#include <nordic/nrf52832.dtsi>
```
Comment (galak) [inline boards/arm/nrf52832_mdk/nrf52832_mdk.dts]: change to include <nordic/nrf52832_qfaa.dtsi>

**Review by galak — CHANGES_REQUESTED** [review]
A slew of dts fixups, also board docs are missing.

**Review by galak — COMMENTED** [review]
Please merge commits down so its a commit for each board, and a commit to the blink sample

File: soc/arm/nordic_nrf/nrf52/dts_fixup.h (hunk 5 above)
Comment (jpsaman) [reply soc/arm/nordic_nrf/nrf52/dts_fixup.h]: I am new to DTS so, would appreciate pointers.

File: boards/arm/nrf52832_mdk/board.cmake [hunk 11]
```diff
@@ -0,0 +1,2 @@
+board_runner_args(pyocd "--target=nrf52")
+include(${ZEPHYR_BASE}/boards/common/pyocd.board.cmake)
```
Comment (ioannisg) [inline boards/arm/nrf52832_mdk/board.cmake]: Github shows an invalid return/EOL

File: soc/arm/nordic_nrf/nrf52/dts_fixup.h (hunk 5 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf52/dts_fixup.h]: Please, check how the pin configuration is provided in dts files in other nRF52-based Board definitions, as well as the dtsi file for nRF52 SOCs.

**Review by ioannisg — CHANGES_REQUESTED** [review]

File: dts/arm/nordic/nrf52840.dtsi (hunk 7 above)
Comment (jpsaman) [reply dts/arm/nordic/nrf52840.dtsi]: hmm, that does not work

File: boards/arm/nrf52832_mdk/nrf52832_mdk.dts (hunk 10 above)
Comment (jpsaman) [reply boards/arm/nrf52832_mdk/nrf52832_mdk.dts]: ok

File: samples/basic/blink_led/src/main.c (hunk 6 above)
Comment (jpsaman) [reply samples/basic/blink_led/src/main.c]: ok

File: boards/arm/nrf52840_mdk/nrf52840_mdk.dts (hunk 8 above)
Comment (jpsaman) [reply boards/arm/nrf52840_mdk/nrf52840_mdk.dts]: pk

File: soc/arm/nordic_nrf/nrf52/dts_fixup.h (hunk 5 above)
Comment (galak) [reply soc/arm/nordic_nrf/nrf52/dts_fixup.h]: @jpsaman as @ioannisg, just take a look at examples in boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts

Here's the commit that adds uart pins if you want more detail:

commit 7c2197f2d84be4b8728e6480a1c0e6669baff171
Author: Paweł Zadrożniak <pawel.zadrozniak@nordicsemi.no>
Date:   Tue Aug 28 16:24:22 2018 +0200

    boards: nrf: uart: Moved UART pin configuration to DTS (nRF boards)

File: dts/arm/nordic/nrf52840.dtsi (hunk 7 above)
Comment (galak) [reply dts/arm/nordic/nrf52840.dtsi]: > hmm, that does not work

What happens?

File: boards/arm/nrf52832_mdk/Kconfig.defconfig (hunk 1 above)
Comment (jpsaman) [reply boards/arm/nrf52832_mdk/Kconfig.defconfig]: ok

File: boards/arm/nrf52840_mdk/Kconfig.defconfig (hunk 3 above)
Comment (jpsaman) [reply boards/arm/nrf52840_mdk/Kconfig.defconfig]: ok

**Comment by ioannisg** [comment]
Also, please, consult the contribution guidelines: they include a description of how the commit messages should be formatted.

https://docs.zephyrproject.org/latest/contribute/contribute_guidelines.html

File: boards/arm/nrf52840_mdk/board.h (hunk 4 above)
Comment (jpsaman) [reply boards/arm/nrf52840_mdk/board.h]: ok

File: boards/arm/nrf52832_mdk/board.h (hunk 2 above)
Comment (jpsaman) [reply boards/arm/nrf52832_mdk/board.h]: ok

File: samples/basic/blink_led/src/main.c (hunk 6 above)
Comment (jpsaman) [reply samples/basic/blink_led/src/main.c]: ok

File: boards/arm/nrf52840_mdk/nrf52840_mdk.dts (hunk 8 above)
Comment (jpsaman) [reply boards/arm/nrf52840_mdk/nrf52840_mdk.dts]: ok

File: dts/arm/nordic/nrf52840.dtsi (hunk 7 above)
Comment (jpsaman) [reply dts/arm/nordic/nrf52840.dtsi]: ok

**Comment by carlescufi** [comment]
@jpsaman here is my suggestion to you:

1) First start with a fresh zephyr clone, pointing to the latest master. You should use your own fork, just like you have it now.
2) Add a remote to the makerdiary zephyr fork (`git remote add makerdiary <url>`)
3) Fetch the changes from makerdiary: `git fetch makerdiary`
4) Cherry-pick the commits from the makerdiary into your fork in master by using `git cherry-pick <SHAs>`. You can get the SHAs by doing `git log makerdiary/master`
5) Squash and fix all the issues and organize the changes into logical commits. You need to use `git rebase -i` to achieve that.
6) `git push -f origin`

File: boards/arm/nrf52832_mdk/nrf52832_mdk.dts (hunk 10 above)
Comment (jpsaman) [reply boards/arm/nrf52832_mdk/nrf52832_mdk.dts]: ok

File: boards/arm/nrf52832_mdk/board.cmake (hunk 11 above)
Comment (jpsaman) [reply boards/arm/nrf52832_mdk/board.cmake]: probably needs dos2unix

**Review by galak — CHANGES_REQUESTED** [review]
Looks a lot better, thanks for making the changes.

There looks like some white space vs tab issues in your changes. (ie use of 8 spaces vs a tab) in the dts{i} files, can you fix that up.

**Comment by jpsaman** [comment]
Is there anything more I need to do/fix ?

File: boards/arm/nrf52840_mdk/nrf52840_mdk_defconfig [hunk 12]
```diff
@@ -0,0 +1,25 @@
+CONFIG_ARM=y
+CONFIG_SOC_FAMILY_NRF=y
+CONFIG_SOC_SERIES_NRF52X=y
+CONFIG_SOC_NRF52840_QIAA=y
+CONFIG_BOARD_NRF52840_MDK=y
+
+# Enable MPU
+CONFIG_ARM_MPU=y
+CONFIG_ARM_MPU_NRF52X=y
```
Comment (ioannisg) [inline boards/arm/nrf52840_mdk/nrf52840_mdk_defconfig]: Here I think you can simply remove line 9

**Review by ioannisg — CHANGES_REQUESTED** [review]

File: boards/arm/nrf52840_mdk/nrf52840_mdk_defconfig (hunk 12 above)
Comment (jpsaman) [reply boards/arm/nrf52840_mdk/nrf52840_mdk_defconfig]: seems to work

**Comment by jpsaman** [comment]
How should I solve this error from shippable?

author email (Zelin <zelin@makerdiary.com>) does not follow the syntax: First Last <email>.

**Review by carlescufi — APPROVED** [review]
Looks good to me now. Thanks @jpsaman

**Review by ioannisg — APPROVED** [review]

**Comment by caizelin** [comment]
@jpsaman 
The blink_led sample does not work correctly. 
`CONFIG_PWM_NRF5_SW_0_CLOCK_PRESCALER=9` should be added.

**Comment by jpsaman** [comment]
@caizelin should that be done for nRF52832-mdk too?

**Comment by jpsaman** [comment]
@caizelin can it be merged now? Or is there something else that needs updating?

**Comment by carlescufi** [comment]
@jpsaman we are missing @galak's approval. You can ping on IRC as well.

**Review by galak — APPROVED** [review]

**Pull request merged**